src/jit/flowgraph.cpp

```diff
@@ -8241,90 +8241,381 @@ void Compiler::fgAddInternal()
+                    comp->lvaSetStruct(returnLocalNum, comp->info.compMethodInfo->args.retTypeClass, true);
+                    returnLocalDsc.lvIsMultiRegRet = true;
+                }
+                else
+                {
+                    assert(!"unreached");
+                }
+
+                if (varTypeIsFloating(returnLocalDsc.lvType))
+                {
+                    comp->compFloatingPointUsed = true;
+                }
+
+                if (!varTypeIsFloating(comp->info.compRetType))
+                {
+                    returnLocalDsc.setPrefReg(REG_INTRET, comp);
+                }
+#ifdef REG_FLOATRET
+                else
+                {
+                    returnLocalDsc.setPrefReg(REG_FLOATRET, comp);
+                }
+#endif
+
+#ifdef DEBUG
+                // This temporary should not be converted to a double in stress mode,
+                // because we introduce assigns to it after the stress conversion
+                returnLocalDsc.lvKeepType = 1;
+#endif
+
+                GenTreePtr retTemp = comp->gtNewLclvNode(returnLocalNum, returnLocalDsc.TypeGet());
+
+                // make sure copy prop ignores this node (make sure it always does a reload from the temp).
+                retTemp->gtFlags |= GTF_DONT_CSE;
+                returnExpr = comp->gtNewOperNode(GT_RETURN, retTemp->gtType, retTemp);
+            }
+            else
+            {
+                // return void
+                noway_assert(comp->info.compRetType == TYP_VOID || varTypeIsStruct(comp->info.compRetType));
+                comp->genReturnLocal = BAD_VAR_NUM;
+
+                returnExpr = new (comp, GT_RETURN) GenTreeOp(GT_RETURN, TYP_VOID);
             }
+
+            // Add 'return' expression to the return block
+            comp->fgInsertStmtAtEnd(newReturnBB, returnExpr);
+            // Flag that this 'return' was generated by return merging so that subsequent
+            // return block morhping will know to leave it alone.
+            returnExpr->gtFlags |= GTF_RET_MERGED;
+
+#ifdef DEBUG
+            if (comp->verbose)
+            {
+                printf("\nmergeReturns statement tree ");
+                printTreeID(returnExpr);
+                printf(" added to genReturnBB %s\n", newReturnBB->dspToString());
+                comp->gtDispTree(returnExpr);
+                printf("\n");
+            }
+#endif
+            assert(index < maxReturns);
+            returnBlocks[index] = newReturnBB;
+            return newReturnBB;
         }
-    }
+
+        // Merge: Find or create an appropriate merged return block for the given input `returnBlock`,
+        // searching up to `searchLimit` for already-created constant return blocks if it returns a
+        // constant.  If a constant return is used, rewrite `returnBlock` to jump to it, else set
+        // `genReturnBB`/`genReturnLocal` so the rewrite can happen at Morph.
+        BasicBlock* Merge(BasicBlock* returnBlock, unsigned searchLimit)
```

Comment:
Maybe add a comment here about what it means when `returnBlock` is `nullptr`.

Also should this method (and others in the class) have the "standard" jit method header comments?

Reply:
Sure; done.